src/geode/basic/database.cpp

```diff
@@ -0,0 +1,345 @@
+                return do_register_data( std::move( data ) );
+            }
+            return do_register_data( std::move( data ) );
+        }
+
+        const std::unique_ptr< Identifier >& do_register_data(
+            std::unique_ptr< Identifier >&& data )
+        {
+            const auto id = data->id();
+            auto new_storage = std::make_shared< Storage >( std::move( data ) );
+            return storage_.emplace( id, std::move( new_storage ) )
+                .first->second->data();
+        }
+
+        void save_data( const std::unique_ptr< Identifier >& data ) const
+        {
+            const auto filename =
+                absl::StrCat( directory_, "/", data->id().string() );
+            std::ofstream file{ filename, std::ofstream::binary };
+            TContext context;
+            for( const auto& serializer : serializers_ )
+            {
+                serializer( std::get< 0 >( context ) );
+            }
+            Serializer archive{ context, file };
+            archive.ext( data, bitsery::ext::StdSmartPtr{} );
+            archive.adapter().flush();
+            OPENGEODE_EXCEPTION( std::get< 1 >( context ).isValid(),
+                "[Database::save_data] Error while writing file: ", filename );
+        }
+
+        std::unique_ptr< Identifier > load_data( const uuid& id ) const
+        {
+            const auto filename = absl::StrCat( directory_, "/", id.string() );
+            std::ifstream file{ filename, std::ifstream::binary };
+            OPENGEODE_EXCEPTION(
+                file, "[Database::load_data] Failed to open file: ", filename );
+            TContext context{};
+            for( const auto& deserializer : deserializers_ )
+            {
+                deserializer( std::get< 0 >( context ) );
+            }
+            Deserializer archive{ context, file };
+            std::unique_ptr< Identifier > data;
+            archive.ext( data, bitsery::ext::StdSmartPtr{} );
+            const auto& adapter = archive.adapter();
+            OPENGEODE_EXCEPTION(
+                adapter.error() == bitsery::ReaderError::NoError
+                    && adapter.isCompletedSuccessfully()
+                    && std::get< 1 >( context ).isValid(),
+                "[Database::load_data] Error while reading file: ", filename );
+            return data;
+        }
+
+    private:
+        std::string directory_;
+        mutable absl::flat_hash_map< uuid, std::shared_ptr< Storage > >
+            storage_;
+        std::vector< serializer_function > serializers_;
+        std::vector< serializer_function > deserializers_;
+    };
+
+    class Database::Data::Impl
+    {
+    public:
+        Impl( std::shared_ptr< Storage > storage )
+            : storage_{ std::move( storage ) }
+        {
+            storage_->new_data_reference();
+        }
+
+        ~Impl()
+        {
+            storage_->delete_data_reference();
+        }
+
+        const Identifier& data() const
```

Comment:
There is a potential bug for "basic" dev. 

If you take the reference 'A,' pass it to another class, and then destroy Data (which was temporarily holding 'A'). This is a lifetime issue for the reference because destroying the Data will trigger the delete_data_reference which implies this data is no longer used. So it could be released from memory by the database at some point. It does not matter if the reference was given to another class.

A way to avoid people to use the design wrong is to force the dev to manually call delete_data_reference. But I find this solution very verbose and constraining... Know your C++ better and add a bug warning in the doc?

Reply:
This returns a const ref, so I'd say the dev should know it mustn't destroy Data, or else the ref points to nothing (If I understand this issue correctly).

Reply:
You at least add a warning message into the docs.